src/threading/thread.cpp

```diff
@@ -0,0 +1,350 @@
+#elif defined(_MSC_VER)
+	struct THREADNAME_INFO {
+		DWORD dwType; // Must be 0x1000
+		LPCSTR szName; // Pointer to name (in user addr space)
+		DWORD dwThreadID; // Thread ID (-1=caller thread)
+		DWORD dwFlags; // Reserved for future use, must be zero
+	};
+#endif
+
+// For bindToProcessor
+#if __FreeBSD_version >= 702106
+	typedef cpuset_t cpu_set_t;
+#elif defined(__linux) || defined(linux)
+	#include <sched.h>
+#elif defined(__sun) || defined(sun)
+	#include <sys/types.h>
+	#include <sys/processor.h>
+	#include <sys/procset.h>
+#elif defined(_AIX)
+	#include <sys/processor.h>
+#elif defined(__APPLE__)
+	#include <mach/mach_init.h>
+	#include <mach/thread_policy.h>
+#endif
+
+
+Thread::Thread(const std::string &name) :
+	name(name),
+	retval(NULL),
+	request_stop(false),
+	running(false)
+#if __cplusplus >= 201103L
+	, thread(NULL)
+#elif !defined(_WIN32)
+	, started(false)
+#endif
+{}
+
+
+void Thread::wait()
+{
+#if __cplusplus >= 201103L
+	if (!thread || !thread->joinable()) return;
+	thread->join();
+#else
+# ifdef _WIN32
+	if (!running) return;
+	WaitForSingleObject(thread, INFINITE);
+# else // pthread
+	void *status;
+	if (!started) return;
+	int ret = pthread_join(thread, &status);
+	assert(!ret);
+	UNUSED(ret);
+	started = false;
+# endif
+#endif
+}
+
+
+bool Thread::start()
+{
+	if (running) return false;
+	request_stop = false;
+
+#if __cplusplus >= 201103L
+	MutexAutoLock l(continue_mutex);
+	thread = new std::thread(theThread, this);
+#else
+# ifdef _WIN32
+	MutexAutoLock l(continue_mutex);
+#  ifdef _WIN32_WCE
+	thread = CreateThread(NULL, 0, theThread, this, 0, &thread_id);
+#  else
+	thread = (HANDLE) _beginthreadex(NULL, 0, theThread, this, 0, &thread_id);
+#  endif
+	if (!thread) return false;
+# else
+	int status;
+
+	pthread_attr_t attr;
+	pthread_attr_init(&attr);
+
+	MutexAutoLock l(continue_mutex);
+
+	status = pthread_create(&thread, &attr, theThread, this);
+	pthread_attr_destroy(&attr);
+
+	if (status) return false;
+# endif
+#endif
+
+#if __cplusplus < 201103L
+	// Wait until running
+	while (!running) {
+# ifdef _WIN32
+		Sleep(1);
+	}
+# else
+		struct timespec req, rem;
+		req.tv_sec = 0;
+		req.tv_nsec = 1000000;
+		nanosleep(&req, &rem);
+	}
+	started = true;
+# endif
+#endif
+	return true;
+}
+
+
+bool Thread::kill()
+{
+#ifdef _WIN32
+	if (!running) return false;
+	TerminateThread(THREAD_HANDLE, 0);
+	CloseHandle(THREAD_HANDLE);
+#else
+	if (!running) {
+		wait();
+		return false;
+	}
+# ifdef __ANDROID__
+	pthread_kill(THREAD_HANDLE, SIGKILL);
+# else
+	pthread_cancel(THREAD_HANDLE);
+# endif
+	wait();
+#endif
+#if __cplusplus >= 201103L
+	delete thread;
+#endif
+	running = false;
+	return true;
+}
+
+
+bool Thread::isSameThread()
+{
+#if __cplusplus >= 201103L
+	return thread->get_id() == std::this_thread::get_id();
+#else
+# ifdef _WIN32
+	return GetCurrentThreadId() == thread_id;
+# else
+	return pthread_equal(pthread_self(), thread);
+# endif
+#endif
+}
+
+
+#if __cplusplus >= 201103L
```

Comment:
I am legitimately confused by this set of #ifdefs.  Can it be cleaned up at all??